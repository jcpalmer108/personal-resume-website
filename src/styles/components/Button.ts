import styled from "styled-components";
import { desktopBreakpoint, redstone, tabletBreakpoint } from "../variables";

export const Link = styled.a`
  width: 186px;
  height: 55px;
  display: flex;
  flex-direction: row;
  align-items: center;
  column-gap: 10px;
  background-color: ${redstone};
  justify-content: center;
  text-decoration: none;
  color: white;
  font-family: Gilroy-Medium;
  font-size: 12px;
  transition: all 500ms ease;

  @media (min-width: ${tabletBreakpoint}px) {
    width: 214px;
    height: 60px;
    font-size: 14px;
    border: 1px solid transparent;
    transition: all 500ms ease;

    &:hover {
      color: ${redstone};
      background: white;
      transition: all 500ms ease;
      font-size: 16px;
      border: ${redstone} 1px solid;

      img {
        filter: invert(59%) sepia(50%) saturate(1158%) hue-rotate(312deg)
          brightness(92%) contrast(95%);
        transition: all 500ms ease;
      }
    }
  }

  @media (min-width: ${desktopBreakpoint}px) {
    width: 218px;
    height: 64px;
  }
`;

export const ArrowRight = styled.img`
  width: 22px;
  padding-bottom: 3px;
`;
