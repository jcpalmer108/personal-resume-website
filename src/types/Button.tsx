export type ButtonProps = {
  label: string;
  url?: string;
  onClick?: Function;
};
