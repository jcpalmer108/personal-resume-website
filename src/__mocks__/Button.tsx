import React from "react";
import { ButtonProps } from "../types/Button";

export function Button({ label, url }: ButtonProps) {
  return (
    <div>
      {label}
      {url}
    </div>
  );
}
