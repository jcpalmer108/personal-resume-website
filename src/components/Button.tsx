import arrow from "../assets/images/icons/arrow-right.svg";
import { ArrowRight, Link } from "../styles/components/Button";
import { ButtonProps } from "../types/Button";

export default function Button({ label, url, onClick }: ButtonProps ) {
  if(!url && !onClick) {
    return null;
  }

  return (
    <div data-testid="Button" onClick={() => { if(onClick) onClick() }}>
      <Link href={url}>
        <span>{label.toUpperCase()}</span>
        <ArrowRight src={arrow} alt="arrow" />
      </Link>
    </div>
  )
}