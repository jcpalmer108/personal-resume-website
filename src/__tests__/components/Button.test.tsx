import { render, screen, fireEvent } from "@testing-library/react";
import Button from "../../components/Button"

// params needed for component
const params = {
  required: {
    label: 'test'
  },
  optional: {
    url: 'www.google.com',
    onClick: jest.fn()
  }
}

describe('Button', () => {
  test('renders if required and all optional params are passed in', () => {
    // given
    const { required, optional } = params;
    const upperCaseLabel = required.label.toUpperCase();
    render(<Button label={required.label} url={optional.url} onClick={optional.onClick}/>)

    // then
    expect(screen.getByTestId("Button")).toMatchSnapshot()
    expect(screen.getByRole('link', { name: `${upperCaseLabel} arrow` })).toHaveAttribute('href', optional.url)
    expect(screen.getByTestId("Button")).toHaveTextContent(upperCaseLabel)
    expect(screen.getByRole('img', { name: 'arrow' })).toHaveAttribute('src', 'arrow-right.svg')
  })

  test('renders if label and url params are passed in', () => {
    // given
    const { required, optional } = params;
    const upperCaseLabel = required.label.toUpperCase();
    render(<Button label={required.label} url={optional.url}/>)

    // then
    expect(screen.getByTestId("Button")).toMatchSnapshot()
    expect(screen.getByRole('link', { name: `${upperCaseLabel} arrow` })).toHaveAttribute('href', optional.url)
    expect(screen.getByTestId("Button")).toHaveTextContent(upperCaseLabel)
    expect(screen.getByRole('img', { name: 'arrow' })).toHaveAttribute('src', 'arrow-right.svg')
  })

  test('renders if label and onClick params are passed in', () => {
    // given
    const { required, optional } = params;
    const upperCaseLabel = required.label.toUpperCase();
    render(<Button label={required.label} onClick={optional.onClick}/>)

    // then
    expect(screen.getByTestId("Button")).toMatchSnapshot()
    expect(screen.queryByRole('link', { name: `${upperCaseLabel} arrow` })).toBeFalsy()
    expect(screen.getByTestId("Button")).toHaveTextContent(upperCaseLabel)
    expect(screen.getByRole('img', { name: 'arrow' })).toHaveAttribute('src', 'arrow-right.svg')
  })

  test('does not render if only label is passed in', () => {
    // given
    const { required } = params;
    render(<Button label={required.label} />)

    // then
    expect(screen.queryByTestId("Button")).toBeFalsy()
  })

  test('executes onClick function when component is clicked', () => {
    // given
    const { required, optional } = params;
    render(<Button label={required.label} onClick={optional.onClick}/>)

    // when
    fireEvent.click(screen.getByTestId("Button"))

    // then
    expect(optional.onClick).toHaveBeenCalledTimes(1)
  })
})
