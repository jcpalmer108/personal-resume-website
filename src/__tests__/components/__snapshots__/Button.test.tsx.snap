// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders if only required params are passed in 1`] = `
<a
  class="sc-bczRLJ iNjNxX"
  data-testid="Button"
  href="https://www.google.com"
>
  <span>
    TEST
  </span>
  <img
    alt="arrow"
    class="sc-gsnTZi gbyMYH"
    src="arrow-right.svg"
  />
</a>
`;
